### ISSUE TITLE
Just give up and provide the `new` keyword

### ISSUE THREAD

**Issue by masak** [masak/007#147]
From a part of #32 that has been implemented already:

> We could have used the `new` keyword in the construction syntax, but I don't feel that it adds anything.

I now think I was wrong, and `new` does add something. :(

Here's the reason we should have `new`. Quick, what does the following code do?

    if op.operand ~~ Q::Identifier {
        my c = eval(op.operand);
        my arguments = op.argumentlist.arguments.map(eval);
        return call(c, arguments);
    }

Right, you saw the trap from miles away. It's a parse error because the `if` statement doesn't have a block, and because `my c` isn't an allowable inside of an object literal.

Alternatively, you didn't see the trap and expected something like the above to work. (Like me.)

Currently, the "correct" way to write something like the above is this:

    if (op.operand ~~ Q::Identifier) {
        # code inside unchanged
    }

These parentheses are very traditional and no-one has any trouble reading them, but it's a sad state of affairs that they have to be added in order for the program to Do The Right Thing. The rule the programmer would have to remember here is something like "if the `if` statement expression ends with `~~ SomeType`, then add parentheses around the whole expression".

This already happened to me with `None` and was fixed in 6f7c0ded12796fa8ec3325eb90f2257c81373db7. But then it happened with other types too, and I realized it's a general problem.

Basically, `new` is a small two-term slang that fixes this. `new` conveys "here comes an (optional) type name, and then an object literal".

Oh, that reminds me. I think we should require `new` on everything except implicit-`Object` object literals. So both of the below syntaxes would be fine:

    my o1 = {};
    my o2 = new {};

Though it would be *consistent*, I don't feel we need to do the same for arrays. That is `[]` would still be correct, but `new []` wouldn't.

**Comment by vendethiel** [comment]
so, is `{}` an empty object or a standalone block?

**Comment by masak** [comment]
Depends if you're at the start of a statement or not.

    $ bin/007 -e='say(type(quasi { {} })); say(type(quasi { ({}) }))'
    <type Q::Block>
    <type Q::Term::Object>

This overloading is quite common. (Same thing in Perl 5, Perl 6, and JavaScript, to name a few.)

**Comment by vendethiel** [comment]
sure – and they're known to create a few issues ... Or at least some laughter. i.e. `{} + []`.

**Comment by masak** [comment]
Aye.

We don't have that particular one, thanks to Perl 6's "semicolon or newline after `}`" rule.

    $ bin/007 -e='{} + {}'
    Could not parse program

Error message could be better, but the reason the program doesn't parse is that rule. Other languages in the C syntax lineage, including Perl 5, are often too lenient, and auto-close the statement and start a new one.

**Comment by vendethiel** [comment]
That seems fine to me, then!
